### PR TITLE
Add CRD and reconciler for streaming/provider

### PR DESCRIPTION
**Pull request by ericbottard**
Upon creation of a "provider", which can have a type and config (currently
only supports Kafka), will spin up an instance of liiklus configured to
talk to the broker (Kafka atm) and a *-provisioner (again, currently
only supports kafka).

### CODE REVIEW

File: pkg/controllers/streaming/provider_controller.go [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+	return deployment, nil
+}
+
+func (r *ProviderReconciler) constructProvisionerDeploymentForProvider(provider *streamingv1alpha1.Provider) (*appsv1.Deployment, error) {
+	labels := r.constructProvisionerLabelsForProvider(provider)
+
+	switch provider.Spec.BrokerType {
+	case streamingv1alpha1.KafkaBroker:
+		env := []corev1.EnvVar{
+			{Name: "GATEWAY", Value: provider.Status.LiiklusServiceName + ":6565"}, // TODO lookup svc
```
[the remark this answers is not in the document]
Comment (ericbottard) [reply pkg/controllers/streaming/provider_controller.go]: I left the lookup comment for not hardcoding the port, not for resolving the host. For pods living in a given namespace, a simple service name resolves correctly (in the current namespace)

File: pkg/controllers/streaming/provider_controller.go [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+			{Name: "BROKER", Value: provider.Spec.Config["bootstrapServers"]},
+		}
+		deployment := &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels:       labels,
+				Annotations:  make(map[string]string),
+				GenerateName: fmt.Sprintf("%s-provisioner-", provider.Name),
```
[the remark this answers is not in the document]
Comment (ericbottard) [reply pkg/controllers/streaming/provider_controller.go]: Fine, but ... why?
This is to limit the risk of clashes with other (simple) names which may happen to exist. What's the downside of `generateName` that I'm not aware of?

File: go.mod [hunk 3]
```diff
@@ -4,9 +4,18 @@ go 1.13
 
 require (
 	github.com/go-logr/logr v0.1.0
+	github.com/gogo/protobuf v1.2.1 // indirect
 	github.com/google/go-cmp v0.3.1
+	github.com/json-iterator/go v1.1.6 // indirect
+	github.com/modern-go/reflect2 v1.0.1 // indirect
 	github.com/onsi/ginkgo v1.8.0
 	github.com/onsi/gomega v1.5.0
+	github.com/pkg/errors v0.8.1
```
[the remark this answers is not in the document]
Comment (ericbottard) [reply go.mod]: Meant to use `errors` from stdlib

File: pkg/controllers/streaming/provider_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (scothis) [reply pkg/controllers/streaming/provider_controller.go]: If (for whatever reason) the status update fails after the resource is created, the child resource will be orphaned. The next time through the reconciler, the controller will think the child resource doesn't exist and will try to create it. Repeat that cycle a few hundred times and your cluster is out of capacity.

In order to use generate name, the detection of the current resource needs to be much more robust. If the name doesn't exist, then switch to lookup by a label selector and then compare the owner reference before adopting the resource.

[the remark this answers is not in the document]
Comment (ericbottard) [reply pkg/controllers/streaming/provider_controller.go]: 👍

File: cmd/managers/streaming/main.go [hunk 4]
```diff
@@ -57,13 +60,25 @@ func main() {
...
 		LeaderElection:     enableLeaderElection,
+		SyncPeriod:         &syncPeriod,
 	})
 	if err != nil {
 		setupLog.Error(err, "unable to start manager")
 		os.Exit(1)
 	}
 
 	streamControllerLogger := ctrl.Log.WithName("controllers").WithName("Stream")
+	namespace := os.Getenv("SYSTEM_NAMESPACE")
```
Comment (scothis) [inline cmd/managers/streaming/main.go]: nit: move up to line 39

File: config/build/manager/kustomization.yaml [hunk 5]
```diff
@@ -1,2 +1,8 @@
 resources:
 - manager.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Comment (scothis) [inline config/build/manager/kustomization.yaml]: did we figure out where these are coming from? They are impacting every runtime.

File: go.mod [hunk 6]
```diff
@@ -4,9 +4,17 @@ go 1.13
 
 require (
 	github.com/go-logr/logr v0.1.0
+	github.com/gogo/protobuf v1.2.1 // indirect
```
Comment (scothis) [inline go.mod]: are these needed? If I manually remove them they don't come back.

File: pkg/controllers/streaming/config.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package streaming
+
+const (
+	kustomizePrefix     = "riff-streaming"              // kustomize adds this prefix to all our resource names
+	providerImages      = kustomizePrefix + "-provider" // contains image names for the provider
+	liiklusImageKey     = "liiklus"
```
Comment (scothis) [inline pkg/controllers/streaming/config.go]: Let's make these keys more obvious that they represent images. That will make detection with our scanner a bit easier.

```suggestion
	liiklusImageKey     = "liiklusImage"
```

Even more scanner friendly would be a configmap like:

```
data:
  liiklus: |
    image: ...
```

...and then marshal the liiklus string into a map to get the image value. I don't think that's worthwhile at this point, but it's how Knative approaches their configs.

File: go.mod (hunk 6 above)
Comment (ericbottard) [reply go.mod]:
```
go mod why -m github.com/gogo/protobuf
# github.com/gogo/protobuf
github.com/projectriff/system/cmd/managers/build
k8s.io/apimachinery/pkg/runtime
github.com/gogo/protobuf/proto
```

Leaving this as is for now, we can revisit if this seems wrong

**Comment by ericbottard** [comment]
Merged as rebased/squashed f9d0576

Will open a different PR for the rename

**Pull request closed without merging**